### ISSUE TITLE
twisted.conch.checkers.SSHPublicKeyDatabase can't find user's private key

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3984]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| esteve reported|
|-|-|
|Trac ID|trac#3984|
|Type|defect|
|Created|2009-08-27 15:34:31Z|
|Branch|https://github.com/twisted/twisted/tree/key-database-3984|

User's private key can't be found because SSHPublicKeyDatabase#checkKey doesn't expand user's home directory properly. Adding a print statement or a call to log.msg shows that, when user foo tries SSH to a Conch service, SSHPublicKeyDatabase looks for its private key in /home/foo/foo/.ssh

Attachments:

* [3984-sshpublickeydatabase-userdir-lookup.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e7d/e7df733cbacaabbd29e7806c1bbdb30ea8164b63/b0decfcf2303b7698ebef8c725935ffdabd793d9.patch) (3561 bytes) - added by esteve on 2009-08-27 16:06:01Z - 
* [3984-sshpublickeydatabase-userdir-lookup-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e7d/e7df733cbacaabbd29e7806c1bbdb30ea8164b63/d2ced4b19518c62acdd4fe43a4c48d2d1ae57750.patch) (7265 bytes) - added by esteve on 2009-08-31 18:00:31Z - 
* [27989-Fix-key-checker.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e7d/e7df733cbacaabbd29e7806c1bbdb30ea8164b63/54dfaa5d46b0caedc63fedcf032fd3017c3de96c.patch) (3616 bytes) - added by bshi on 2010-01-14 02:54:29Z - Bug fix against SVN 27989

<details><summary>Searchable metadata</summary>

```
trac-id__3984 3984
type__defect defect
reporter__esteve esteve
priority__normal normal
milestone__ 
branch__branches_key_database_3984 branches/key-database-3984
branch_author__bshi bshi
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1251387271000000 1251387271000000
changetime__1263482119000000 1263482119000000
version__None None
owner__ 
cc__bshi
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|esteve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|esteve commented|
|-|-|

This patch changes SSHPublicKeyDatabase#checkKey to use ~user in os.path.expanduser to return, e.g., /home/user, instead of /home/user/user. Tests were also changed to use the full path to the ssh directory ($HOME/.ssh). 

I've also pushed a branch to Launchpad:

https://code.launchpad.net/~esteve/twisted/3984-sshpublickeydatabase-userdir-lookup

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to esteve|
|-|-|

Thanks for noticing this and contributing the patch.

It appears to fix the issue and at least nominally test it, but it seems like it would be very easy to change the code or test a little bit so that it was no longer covered.

  1. Some test's docstring somewhere should talk about how the .ssh directory is located.  The docstring of `SSHPublicKeyDatabase` itself should also be a little more specific.
  1. `MockOS.expanduser` should change so that it gives something different than a static path regardless of its input.  You can currently replace the input to the `expanduser` call in `checkKey` to anything at all (even a non-string) and the test still passes.
  1. `self.sshDir` in the test case should be set up by doing something like `FilePath(self.mktemp()).child(".ssh")`; avoid the use of `os.path` when you can.

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|esteve set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|esteve commented|
|-|-|

Thanks Glyph for you review.

I just uploaded a new version of the patch and pushed its changes to Launchpad. I changed MockOS to return different user home directory depending on the argument passed. I haven't added tests for MockOS#expanduser, because I preferred to know if I was on the right track before going any further.

I also changed the test suite to use FilePath instead of os.path.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to esteve<br>@exarkun set status to `new`|
|-|-|

  1. `MockOS.users` needs to be an instance attribute.  As implemented now, every `MockOS` instance shares a single users dict.  That will have some Consequences. :)
  1. Since `SSHPublicKeyDatabase.checkKey` already uses the `pwd` module, maybe it would be simpler to implement and test a solution based on that.  The `pwd` module will tell you the home directory of a user, and this won't involve string parsing or as much path mangling to fake in the tests.  Plus, there's `twisted.python.fakepwd` to make it extra simple.
  1. In the test setup, if you set `self.mockos.path` to `FilePath(self.mktemp())` first and construct the rest of the paths based on that, you won't need `sshDir.parent().parent().parent()` :)
  1. I'm puzzled by the changes to `MockOS.setuid` and `MockOS.setgid`.  What purpose do they serve?

**Comment by twisted-trac** [comment]
|<img alt="bshi's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|bshi removed owner|
|-|-|

Anyone working on this?  Please review proposed patch to resolve this attached as 27989-Fix-key-checker.patch which incorporates the suggestions in the most recent review.

After skimming the OpenSSH documentation, refactored checkKey() slightly and added a new class method to get a list of authorized_key files.  Idea here is that API users can override for custom configurations.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [27990]) Branching to 'key-database-3984'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Thanks for picking this up bshi.  The patch looks great.  I'm going to make a few minor adjustments and commit it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [27992]) two blank lines between methods; update copyright header; convert todo in docstring to a note about the implementation's behavior; use `FilePath` instead of `os.path`

refs [#3984](https://github.com/twisted/twisted/issues/3984)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [27993]) news

refs [#3984](https://github.com/twisted/twisted/issues/3984)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [27994]) Merge key-database-3984

Author: bshi
Reviewer: exarkun
Fixes: [#3984](https://github.com/twisted/twisted/issues/3984)

Fix `SSHPublicKeyDatabase` so that it looks for authorized_keys files in
the right place.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
